### DISCUSSION TITLE
Modbus driver for Alsavo 1401

### DISCUSSION THREAD

**Discussion by vendersgade** [strandborg/AlsavoCtrl#28] [Q&A]
Dear All,

I created a prototype Modbus RS485 driver in LabVIEW for the 1401 variant using COM2 (3 first pins from the left are A,B and GND) ( One issue I have is that I cant run the compressor below 29 Hz or the fan above 1000 RPM, any help on this would be greatly apriciated).

Initially I tried to create a TCP driver, but never managed to get it to answer to any UPD commands/connection. I assume that the unit must have been in LAN mode, since I was able to control it via the APP, but I am not sure. Hence my question: Is there any specific settings needed to activate LAN mode ? Btw. the wireless LAN card uses aprox. 7-8 watt's and it can be disconnected and the unit runs perfectly withouut, provided it receives periodic RS485 commands on COM2. The internal watchdog will shutdown the unit if there is no communication. 

Strandborg : thanks for sharing your code, couldnt have made mine without inspiration from yours.

**Comment by strandborg** [comment]
Awesome! I'd love to get more information about the MODBUS implementation,
the wifi stack in the device seems to be pretty unreliable. As for fully
manual control, I don't think that's a good idea, as you'd also have to
adjust EEV manually, and calculating that properly might be tricky, the
algorithm isn't publicly available.

**Reply by strandborg** [reply]
@vendersgade Do you have any more information available of the modbus connection?

**Comment by vendersgade** [comment]
I basically reverse engineered the "driver", by first adding a sniffer between the display and the main board (thats also Modbus communication) and then just identifying all the different commands. 

Adjusting the EEV, Fan speed and compressor rate manually (or programatically in manual mode of the heat pump) is not that difficult really.  Despite the unit being in "manual" there are not endless amounts of freedoms, it will perform a shutdown if the pressure is too high etc. so even in manual, the safety is implemented. The native control algorithm for the unit does not seem to be optimal, thats the hole idea behind my project. For instance, de-icing can be improved a lot. 

I have been able to run the modbus communication without any glitches for days at a time, but the display needs to be disconnected (there can only be one master).  Currently i run the code from a PC, but my main goal is to run it from an Arduino or BBB

I'm off for a much needed vacation in Sweden, but I would not mind uploading my code at a later point (it sill needs polishing). But please look back in a few weeks time.

**Reply by stefanh12** [reply]
Could you post your modbus code? Trying to get this setup with esphome (Brilix), sniffing the packets between my screen and the heat pump seems not to be modbus though.

**Comment by vendersgade** [comment]
![image](https://github.com/user-attachments/assets/9793e9bb-59a0-4a4a-ae3c-db841bd041f6)
![image](https://github.com/user-attachments/assets/1d926b48-6a2a-4a93-8f83-3d0a6bd660ea)

**Comment by vendersgade** [comment]
I'll guarantee you that its Modbus RTU (not ASCII)

Here is one of the commands and the responce.
0204 - Read input registers CMD : 
0204 0000 000E 71FD 

and the system responds : 
0204 2800 B900 CE00 CD00 0000 CB00 C600 7800 0000 0000 FA00 0000 0001 9A00 3C00 0000 0006 4300 0000 0000 0071 B1

Thats 100% a Modbus frame stucture. 

You can scan all the registers with any modbus scanner tool. Such as CAS modbus scanner. Pinouts and wire connections are in my first post. 

The code I made is in LabVIEW (graphical). It compiles directly to machine language, so there is no text behind the graphics.

**Reply by stefanh12** [reply]
Thats awsome, I was a bit naive and thought I could capture the frames using a Waveshare tcp rs485 converter, then using a modbus frame analyser, but I just get CRC errors. So in essence im a bit over my head.  I'm glad thought thats its modbus! It would be really great if we could figure out how to get this setup using esphome and a LILYGO T-CAN485. That would be a robust implementation. Could you share your findings on registers, coils and and I might be able to get esphome code I can share with you all.

**Reply by stefanh12** [reply]
I've managed to get a bit longer now using your information! I'm now able to intercept and decode the modbus messages. I've not managed to map them all yet, have you @vendersgade ?

Here is a decoded of the 3 different packages, looks that's all the registers that the display reads and writes too.
[Namnlöst.pdf](https://github.com/user-attachments/files/16906653/Namnlost.pdf)

**Comment by stefanh12** [comment]
Winter is getting close here in Sweden. I've managed to get a prototype working with esphome but I have not been able to figure out that many modbus registers as I would want to. ill link to the code later

**Comment by stefanh12** [comment]
Started to fill my pool after winter now, I've created a esphome version of what I've found out from you all and sniffing the communication and will start testing it during the next weeks.

This has to be the cheapest and most stable way to do this if your using Home Assistant. 

Remove the display, connect a t-can485 instead of the display. There are a lot of registers that I have no clue on what they do but I've added them so it's easier to figure out what they are. I've also not seen the device use the bitmask config but instead registers for each setting.

https://lilygo.cc/products/t-can485
https://github.com/stefanh12/HA/blob/main/brilix-heat-pump.yaml

**Reply by stefanh12** [reply]
Made a lot of changes to the esphome code since the heat pump does not support reading coils and registers. Changed so that write of the coils and registers are done on an interval to handle the issue with esphome not made for this type of application. I can now set temp and turn the unit on /off. Lots of things I still do not know what they are, so if anyone wishes to help it would be great

**Comment by stefanh12** [comment]
Been running this with esphome and home assistant all summer, its really working great

**Comment by parrafin23** [comment]
Tjena Stefan!

I bought the fun'n'swim swimmingpool heatpumpthingy, got it up and running, Lookin at the fan, the big spinny thingy thats inside the metal box.. lookin at the speed of that thing doesent rly blow your socks off now does it, compared to a regular heat pump for your house.. Have you found a way to up the rippms of the fan to have better efficency of the pump? I do have Ha installed and I would like just as much as you to have more control of the device my self. pref. take it out of the cloud if possible. and ye, the pool is filled and the pump is chummin away trying its best to keep the water warm, not easy up here in the north,,

**Reply by stefanh12** [reply]
I have not checked this at all, only running the smart and other modes
